src/msg/xio/XioMessenger.cc

```diff
@@ -0,0 +1,1041 @@
+			     void *conn_user_context)
+{
+  XioConnection* xcon __attribute__((unused)) =
+    static_cast<XioConnection*>(conn_user_context);
+  CephContext *cct = xcon->get_messenger()->cct;
+
+  ldout(cct,25) << "on cancel request: session: " << session << " msg: " << msg
+    << " conn_user_context " << conn_user_context << dendl;
+
+  return 0;
+}
+
+/* free functions */
+static string xio_uri_from_entity(const entity_addr_t& addr, bool want_port)
+{
+  const char *host = NULL;
+  char addr_buf[129];
+
+  switch(addr.addr.ss_family) {
+  case AF_INET:
+    host = inet_ntop(AF_INET, &addr.addr4.sin_addr, addr_buf,
+		     INET_ADDRSTRLEN);
+    break;
+  case AF_INET6:
+    host = inet_ntop(AF_INET6, &addr.addr6.sin6_addr, addr_buf,
+		     INET6_ADDRSTRLEN);
+    break;
+  default:
+    abort();
+    break;
+  };
+
+  /* The following can only succeed if the host is rdma-capable */
+  string xio_uri = "rdma://";
+  xio_uri += host;
+  if (want_port) {
+    xio_uri += ":";
+    xio_uri += boost::lexical_cast<std::string>(addr.get_port());
+  }
+
+  return xio_uri;
+} /* xio_uri_from_entity */
+
+/* XioMessenger */
+XioMessenger::XioMessenger(CephContext *cct, entity_name_t name,
+			   string mname, uint64_t nonce,
+			   DispatchStrategy *ds)
+  : SimplePolicyMessenger(cct, name, mname, nonce),
+    nsessions(0),
+    shutdown_called(false),
+    portals(this, cct->_conf->xio_portal_threads),
+    dispatch_strategy(ds),
+    loop_con(this),
+    special_handling(0),
+    sh_mtx("XioMessenger session mutex"),
+    sh_cond()
+{
+
+  if (cct->_conf->xio_trace_xcon)
+    magic |= MSG_MAGIC_TRACE_XCON;
+
+  XioPool::trace_mempool = (cct->_conf->xio_trace_mempool);
+  XioPool::trace_msgcnt = (cct->_conf->xio_trace_msgcnt);
+
+  /* package init */
+  if (! initialized.read()) {
+
+    mtx.Lock();
+    if (! initialized.read()) {
+
+      xio_init();
+
+      // claim a reference to the first context we see
+      xio_log::context = cct->get();
+
+      int xopt;
+      xopt = xio_log::get_level();
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_LOG_LEVEL,
+		  &xopt, sizeof(xopt));
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_LOG_FN,
+		  (const void*)xio_log::log_dout, sizeof(xio_log_fn));
+
+      xopt = 1;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_DISABLE_HUGETBL,
+		  &xopt, sizeof(xopt));
+
+      xopt = XIO_MSGR_IOVLEN;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_MAX_IN_IOVLEN,
+		  &xopt, sizeof(xopt));
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_MAX_OUT_IOVLEN,
+		  &xopt, sizeof(xopt));
+
+      /* enable flow-control */
+      xopt = 1;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_ENABLE_FLOW_CONTROL,
+                 &xopt, sizeof(xopt));
+
+      /* and set threshold for buffer callouts */
+      xopt = 16384;
+      xio_set_opt(NULL, XIO_OPTLEVEL_ACCELIO, XIO_OPTNAME_MAX_INLINE_DATA,
+                 &xopt, sizeof(xopt));
+      xopt = 216;
```

Comment:
Matt, can you increase the max inline header size supported to atleast 218 bytes? Kernel RBD sends the ceph header as is(without encoding) and contains a 2 byte reserved field that is not handled if max inline header size in just 216.